File: languages/csharp/AppEncryption/AppEncryption/AppEncryptionSessionFactory.cs
using System;
using App.Metrics;
using GoDaddy.Asherah.AppEncryption.Envelope;
using GoDaddy.Asherah.AppEncryption.KeyManagement;
using GoDaddy.Asherah.AppEncryption.Persistence;
using GoDaddy.Asherah.AppEncryption.Util;
using GoDaddy.Asherah.Crypto;
using GoDaddy.Asherah.Crypto.Engine.BouncyCastle;
using GoDaddy.Asherah.Crypto.Keys;
using GoDaddy.Asherah.Logging;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;

namespace GoDaddy.Asherah.AppEncryption
{
    public class AppEncryptionSessionFactory : IDisposable
    {
        private static readonly ILogger Logger = LogManager.CreateLogger<AppEncryptionSessionFactory>();

        private readonly string productId;
        private readonly string systemId;
        private readonly IMetastorePersistence<JObject> metastorePersistence;
        private readonly SecureCryptoKeyDictionaryFactory<DateTimeOffset> secureCryptoKeyDictionaryFactory;
        private readonly SecureCryptoKeyDictionary<DateTimeOffset> systemKeyCache;
        private readonly CryptoPolicy cryptoPolicy;
        private readonly KeyManagementService keyManagementService;

        public AppEncryptionSessionFactory(
            string productId,
            string systemId,
            IMetastorePersistence<JObject> metastorePersistence,
            SecureCryptoKeyDictionaryFactory<DateTimeOffset> secureCryptoKeyDictionaryFactory,
            CryptoPolicy cryptoPolicy,
            KeyManagementService keyManagementService)
        {
            this.productId = productId;
            this.systemId = systemId;
            this.metastorePersistence = metastorePersistence;
            this.secureCryptoKeyDictionaryFactory = secureCryptoKeyDictionaryFactory;
            systemKeyCache = secureCryptoKeyDictionaryFactory.CreateSecureCryptoKeyDictionary();
            this.cryptoPolicy = cryptoPolicy;
            this.keyManagementService = keyManagementService;
        }

        public interface IMetastoreStep
        {
            // Leaving this in here for now for user integration test convenience. Need to add "don't run in prod" checks somehow
            ICryptoPolicyStep WithMemoryPersistence();

            ICryptoPolicyStep WithMetaStorePersistence(IMetastorePersistence<JObject> persistence);
        }

        public interface ICryptoPolicyStep
        {
            IKeyManagementServiceStep WithNeverExpiredCryptoPolicy();

            IKeyManagementServiceStep WithCryptoPolicy(CryptoPolicy cryptoPolicy);
        }

        public interface IKeyManagementServiceStep
        {
            // Leaving this in here for now for user integration test convenience. Need to add "don't run in prod" checks somehow
            IBuildStep WithStaticKeyManagementService(string demoMasterKey);

            IBuildStep WithKeyManagementService(KeyManagementService keyManagementService);
        }

        public interface IBuildStep
        {
            IBuildStep WithMetrics(IMetrics metrics);

            AppEncryptionSessionFactory Build();
        }

        public static IMetastoreStep NewBuilder(string productId, string systemId)
        {
            return new Builder(productId, systemId);
        }

        public virtual void Dispose()
        {
            try
            {
                // only close system key cache since we invoke its creation
                systemKeyCache.Dispose();
            }
            catch (Exception e)
            {
                Logger.LogError(e, "unexpected exception during close");
            }
        }

        public AppEncryption<JObject, byte[]> GetAppEncryptionJson(string partitionId)
        {
            IEnvelopeEncryption<byte[]> envelopeEncryption = GetEnvelopeEncryptionBytes(partitionId);

            return new AppEncryptionJsonImpl<byte[]>(envelopeEncryption);
        }

        public AppEncryption<byte[], byte[]> GetAppEncryptionBytes(string partitionId)
        {
            IEnvelopeEncryption<byte[]> envelopeEncryption = GetEnvelopeEncryptionBytes(partitionId);

            return new AppEncryptionBytesImpl<byte[]>(envelopeEncryption);
        }

        public AppEncryption<JObject, JObject> GetAppEncryptionJsonAsJson(string partitionId)
        {
            IEnvelopeEncryption<JObject> envelopeEncryption = GetEnvelopeEncryptionJson(partitionId);

            return new AppEncryptionJsonImpl<JObject>(envelopeEncryption);
        }

        public AppEncryption<byte[], JObject> GetAppEncryptionBytesAsJson(string partitionId)
        {
            IEnvelopeEncryption<JObject> envelopeEncryption = GetEnvelopeEncryptionJson(partitionId);

            return new AppEncryptionBytesImpl<JObject>(envelopeEncryption);
        }

        internal IEnvelopeEncryption<byte[]> GetEnvelopeEncryptionBytes(string partitionId)
        {
            return new EnvelopeEncryptionBytesImpl(GetEnvelopeEncryptionJson(partitionId));
        }

        internal AppEncryptionPartition GetAppEncryptionPartition(string partitionId)
        {
            return new AppEncryptionPartition(partitionId, systemId, productId);
        }

        private EnvelopeEncryptionJsonImpl GetEnvelopeEncryptionJson(string partitionId)
        {
            AppEncryptionPartition appEncryptionPartition = GetAppEncryptionPartition(partitionId);

            return new EnvelopeEncryptionJsonImpl(
                appEncryptionPartition,
                metastorePersistence,
                systemKeyCache,
                secureCryptoKeyDictionaryFactory,
                new BouncyAes256GcmCrypto(),
                cryptoPolicy,
                keyManagementService);
        }

        private class Builder : IMetastoreStep, ICryptoPolicyStep, IKeyManagementServiceStep, IBuildStep
        {
            private readonly string productId;
            private readonly string systemId;

            private IMetastorePersistence<JObject> metastorePersistence;
            private CryptoPolicy cryptoPolicy;
            private KeyManagementService keyManagementService;
            private IMetrics metrics;

            internal Builder(string productId, string systemId)
            {
                this.productId = productId;
                this.systemId = systemId;
            }

            public ICryptoPolicyStep WithMemoryPersistence()
            {
                metastorePersistence = new MemoryPersistenceImpl<JObject>();
                return this;
            }

            public ICryptoPolicyStep WithMetaStorePersistence(IMetastorePersistence<JObject> persistence)
            {
                metastorePersistence = persistence;
                return this;
            }

            public IKeyManagementServiceStep WithNeverExpiredCryptoPolicy()
            {
                cryptoPolicy = new NeverExpiredCryptoPolicy();
                return this;
            }

            public IKeyManagementServiceStep WithCryptoPolicy(CryptoPolicy policy)
            {
                cryptoPolicy = policy;
                return this;
            }

            public IBuildStep WithStaticKeyManagementService(string demoMasterKey)
            {
                keyManagementService = new StaticKeyManagementServiceImpl(demoMasterKey);
                return this;
            }

            public IBuildStep WithKeyManagementService(KeyManagementService kms)
            {
                keyManagementService = kms;
                return this;
            }

            public IBuildStep WithMetrics(IMetrics metrics)
            {
                this.metrics = metrics;
                return this;
            }

            public AppEncryptionSessionFactory Build()
            {
                // If no metrics provided, we just create a disabled/no-op one
                if (metrics == null)
                {
                    metrics = new MetricsBuilder()
                        .Configuration.Configure(options => options.Enabled = false)
                        .Build();
                }

                MetricsUtil.SetMetricsInstance(metrics);

                return new AppEncryptionSessionFactory(
                    productId,
                    systemId,
                    metastorePersistence,
                    new SecureCryptoKeyDictionaryFactory<DateTimeOffset>(cryptoPolicy),
                    cryptoPolicy,
                    keyManagementService);
            }
        }
    }
}

File: languages/java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/AppEncryptionSessionFactory.java
package com.godaddy.asherah.appencryption;

import java.time.Instant;

import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryption;
import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryptionBytesImpl;
import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryptionJsonImpl;
import com.godaddy.asherah.appencryption.keymanagement.KeyManagementService;
import com.godaddy.asherah.appencryption.keymanagement.StaticKeyManagementServiceImpl;
import com.godaddy.asherah.appencryption.persistence.MemoryPersistenceImpl;
import com.godaddy.asherah.appencryption.persistence.MetastorePersistence;
import com.godaddy.asherah.appencryption.utils.MetricsUtil;
import com.godaddy.asherah.appencryption.utils.SafeAutoCloseable;
import com.godaddy.asherah.crypto.CryptoPolicy;
import com.godaddy.asherah.crypto.NeverExpiredCryptoPolicy;
import com.godaddy.asherah.crypto.engine.bouncycastle.BouncyAes256GcmCrypto;
import com.godaddy.asherah.crypto.keys.SecureCryptoKeyMap;
import com.godaddy.asherah.crypto.keys.SecureCryptoKeyMapFactory;

import io.micrometer.core.instrument.Metrics;
import io.micrometer.core.instrument.config.MeterFilter;

public class AppEncryptionSessionFactory implements SafeAutoCloseable {
  private static final Logger logger = LoggerFactory.getLogger(AppEncryptionSessionFactory.class);

  private final String productId;
  private final String systemId;
  private final MetastorePersistence<JSONObject> metastorePersistence;
  private final SecureCryptoKeyMapFactory<Instant> secureCryptoKeyMapFactory;
  private final SecureCryptoKeyMap<Instant> systemKeyCache;
  private final CryptoPolicy cryptoPolicy;
  private final KeyManagementService keyManagementService;

  public AppEncryptionSessionFactory(
      final String productId,
      final String systemId,
      final MetastorePersistence<JSONObject> metastorePersistence,
      final SecureCryptoKeyMapFactory<Instant> secureCryptoKeyMapFactory,
      final CryptoPolicy cryptoPolicy,
      final KeyManagementService keyManagementService) {
    this.productId = productId;
    this.systemId = systemId;
    this.metastorePersistence = metastorePersistence;
    this.secureCryptoKeyMapFactory = secureCryptoKeyMapFactory;
    this.systemKeyCache = secureCryptoKeyMapFactory.createSecureCryptoKeyMap();
    this.cryptoPolicy = cryptoPolicy;
    this.keyManagementService = keyManagementService;
  }

  public AppEncryption<JSONObject, byte[]> getAppEncryptionJson(final String partitionId) {
    EnvelopeEncryption<byte[]> envelopeEncryption = getEnvelopeEncryptionBytes(partitionId);

    return new AppEncryptionJsonImpl<>(envelopeEncryption);
  }

  public AppEncryption<byte[], byte[]> getAppEncryptionBytes(final String partitionId) {
    EnvelopeEncryption<byte[]> envelopeEncryption = getEnvelopeEncryptionBytes(partitionId);

    return new AppEncryptionBytesImpl<>(envelopeEncryption);
  }

  public AppEncryption<JSONObject, JSONObject> getAppEncryptionJsonAsJson(final String partitionId) {
    EnvelopeEncryption<JSONObject> envelopeEncryption = getEnvelopeEncryptionJson(partitionId);

    return new AppEncryptionJsonImpl<>(envelopeEncryption);
  }

  public AppEncryption<byte[], JSONObject> getAppEncryptionBytesAsJson(final String partitionId) {
    EnvelopeEncryption<JSONObject> envelopeEncryption = getEnvelopeEncryptionJson(partitionId);

    return new AppEncryptionBytesImpl<>(envelopeEncryption);
  }

  EnvelopeEncryption<byte[]> getEnvelopeEncryptionBytes(final String partitionId) {
    return new EnvelopeEncryptionBytesImpl(getEnvelopeEncryptionJson(partitionId));
  }

  private EnvelopeEncryptionJsonImpl getEnvelopeEncryptionJson(final String partitionId) {
    AppEncryptionPartition appEncryptionPartition = getAppEncryptionPartition(partitionId);
    return new EnvelopeEncryptionJsonImpl(
        appEncryptionPartition,
        metastorePersistence,
        systemKeyCache,
        secureCryptoKeyMapFactory,
        new BouncyAes256GcmCrypto(),
        cryptoPolicy,
        keyManagementService);
  }

  AppEncryptionPartition getAppEncryptionPartition(final String partitionId) {
    return new AppEncryptionPartition(partitionId, systemId, productId);
  }

  @Override
  public void close() {
    try {
      // only close system key cache since we invoke its creation
      systemKeyCache.close();
    }
    catch (Exception e) {
      logger.error("unexpected exception during close", e);
    }
  }

  public static MetastoreStep newBuilder(final String productId, final String systemId) {
    return new Builder(productId, systemId);
  }

  public static final class Builder implements MetastoreStep, CryptoPolicyStep, KeyManagementServiceStep, BuildStep {
    private final String productId;
    private final String systemId;

    private MetastorePersistence<JSONObject> metastorePersistence;
    private CryptoPolicy cryptoPolicy;
    private KeyManagementService keyManagementService;
    private boolean metricsEnabled = false;

    private Builder(final String productId, final String systemId) {
      this.productId = productId;
      this.systemId = systemId;
    }

    @Override
    public CryptoPolicyStep withMemoryPersistence() {
      this.metastorePersistence = new MemoryPersistenceImpl<>();
      return this;
    }

    @Override
    public CryptoPolicyStep withMetastorePersistence(final MetastorePersistence<JSONObject> persistence) {
      this.metastorePersistence = persistence;
      return this;
    }

    @Override
    public KeyManagementServiceStep withNeverExpiredCryptoPolicy() {
      this.cryptoPolicy = new NeverExpiredCryptoPolicy();
      return this;
    }

    @Override
    public KeyManagementServiceStep withCryptoPolicy(final CryptoPolicy policy) {
      this.cryptoPolicy = policy;
      return this;
    }

    @Override
    public BuildStep withStaticKeyManagementService(final String demoMasterKey) {
      this.keyManagementService = new StaticKeyManagementServiceImpl(demoMasterKey);
      return this;
    }


    @Override
    public BuildStep withKeyManagementService(final KeyManagementService kms) {
      this.keyManagementService = kms;
      return this;
    }

    @Override
    public BuildStep withMetricsEnabled() {
      this.metricsEnabled = true;
      return this;
    }

    @Override
    public AppEncryptionSessionFactory build() {
      if (!metricsEnabled) {
        // Deny takes precedence in the filtering logic, so we deny if they didn't explicitly enable metrics
        Metrics.globalRegistry.config().meterFilter(MeterFilter.denyNameStartsWith(MetricsUtil.AEL_METRICS_PREFIX));
      }

      return new AppEncryptionSessionFactory(productId, systemId, metastorePersistence,
          new SecureCryptoKeyMapFactory<>(cryptoPolicy), cryptoPolicy, keyManagementService);
    }
  }

  public interface MetastoreStep {
    // Leaving this here for now for user integration test convenience. Need to add "don't run in prod" checks somehow
    CryptoPolicyStep withMemoryPersistence();

    CryptoPolicyStep withMetastorePersistence(MetastorePersistence<JSONObject> persistence);
  }

  public interface CryptoPolicyStep {
    KeyManagementServiceStep withNeverExpiredCryptoPolicy();

    KeyManagementServiceStep withCryptoPolicy(CryptoPolicy policy);
  }

  public interface KeyManagementServiceStep {
    // Leaving this here for now for user integration test convenience. Need to add "don't run in prod" checks somehow
    BuildStep withStaticKeyManagementService(String demoMasterKey);

    BuildStep withKeyManagementService(KeyManagementService keyManagementService);
  }

  public interface BuildStep {
    BuildStep withMetricsEnabled();

    AppEncryptionSessionFactory build();
  }
}
